### DISCUSSION TITLE
Reconsider when Elasticsearch queries are sent while editing a Filter

### DISCUSSION THREAD

**Discussion by Ultraseamus** [grafana/grafana#61164] [Data Sources]
summary by @gabor : when editing the `query` and the `filters/query` text-fields, they auto-submit the form after a short inactivity-timer. this should change. possibilities:
- A. never auto-submit the form (user has to press the [run query] type buttons) or use keyboard shortcut
- B.  auto-submit when the field loses focus (on the `blur` HTML event)

currently, in other datasources, loki Loki&Prometheus, we do the [A] in explore and [B] in dashboards. there is an effort to standardize the behavior in some way in https://github.com/grafana/grafana/issues/51979

---
**Why is this needed**:

It used to be that a query would run once an element in the Query Editor lost focus. That was not the best experience for new users, and even for experienced users, it is sometimes just a bit of a pain to have to find a free spot to click.

In the new Elasticsearch plugin, queries are sent as you type (with a small delay). This is more intuitive, but it means a user could easily send out dozens of bad queries while they are typing in a medium-length filter. It's not valuable to have a query sent when you are only half-way through typing it, and it results in unnecessary pressure on the Elasticsearch cluster.

**What would you like to be added**:

This should be thought through again. The old method was a bit confusing, but sent fewer impartial queries.

A simple middle ground could be adding a "Run Query" button. Then a user could make alterations to the filters, aggregations, and whatever else, without a potentially slow, expensive queue of queries stacking up. Once they are ready, the new query could be run once they press "Run Query" or when they close the Editor.

I believe this would be an overall improvement over how it is now, and how it has been in the past (especially if you think of making modifications to a query that takes several seconds to return each time). And, I can add one extra example of why this really is valuable.

In one of the Elasticsearch indices my team uses, there is a very long period of data, with multiple possible time fields some at daily intervals, others at monthly intervals (so the configured datasource min interval has to allow 1d). If you have a dashboard scoped to a few years and add this datasource, you have to be very careful with your order of operations. If you select a different time field before you change the interval, you could end up sending a very expensive query, just because changing the time field will run the query before you have the change to change the interval.

That may be a very specific example, but I think the basic idea is generic. Needless queries are also sent out when you add a Terms aggregation, before you can select the field. Then after you select the field, but before you can sort. Then after you sort, but before you can change the sort direction. Every click runs another query.

**Comment by aocenas** [comment]
@torkelo do you think it would make sense to consider unifying how queries are run in explore and in dashboards? I also find the implicit query execution surprising sometimes (especially when accounted for non consistent implementation of the "run on blur") and having just the small refresh button hard to find sometimes. I feel like having consistent "run query button" plus single "run query" shortcut would make sense.

**Comment by matyax** [comment]
Seems related with https://github.com/grafana/grafana/issues/57395

**Comment by matyax** [comment]
See https://github.com/grafana/grafana/issues/51979

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
